dd-trace/src/configuration/configuration.rs

```diff
@@ -156,112 +181,325 @@ impl<T: Deref> Deref for ConfigItemRef<'_, T> {
+    fn get_configuration(&self) -> Configuration {
+        Configuration {
+            name: self.name.to_string(),
+            value: self.value().get_configuration_value(),
+            origin: self.source().into(),
+            config_id: None,
+        }
     }
+}
 
+impl<T: ConfigurationValueProvider> ValueSourceUpdater<T> for ConfigItem<T> {
     /// Sets a value from a specific source
     fn set_value_source(&mut self, value: T, source: ConfigSourceOrigin) {
         match source {
             ConfigSourceOrigin::Code => self.code_value = Some(value),
+            ConfigSourceOrigin::EnvVar => self.env_value = Some(value),
             ConfigSourceOrigin::RemoteConfig => {
-                self.set_rc(value);
+                dd_warn!("Cannot set a value from RC");
             }
-            ConfigSourceOrigin::EnvVar => self.env_value = Some(value),
             ConfigSourceOrigin::Default => {
                 dd_warn!("Cannot set default value after initialization");
             }
         }
     }
+}
 
-    /// Sets the code value (convenience method)
-    fn set_code(&mut self, value: T) {
-        self.code_value = Some(value);
+/// Configuration item that tracks the value of a setting and where it came from
+/// And allows to update the corresponding value with a ConfigSourceOrigin
+#[derive(Debug)]
+struct ConfigItemWithOverride<T: ConfigurationValueProvider> {
+    config_item: ConfigItem<T>,
+    override_value: arc_swap::ArcSwapOption<T>,
+    source_type: ConfigSourceOrigin,
+}
+
+impl<T: Clone + ConfigurationValueProvider> Clone for ConfigItemWithOverride<T> {
+    fn clone(&self) -> Self {
+        Self {
+            config_item: self.config_item.clone(),
+            override_value: arc_swap::ArcSwapOption::new(self.override_value.load_full()),
+            source_type: self.source_type,
+        }
+    }
+}
+
+impl<T: ConfigurationValueProvider + Clone> ConfigItemWithOverride<T> {
+    fn new_code(name: &'static str, default: T) -> Self {
+        Self {
+            config_item: ConfigItem::new(name, default),
+            override_value: arc_swap::ArcSwapOption::const_empty(),
+            source_type: ConfigSourceOrigin::Code,
+        }
+    }
+
+    fn new_rc(name: &'static str, default: T) -> Self {
+        Self {
+            config_item: ConfigItem::new(name, default),
+            override_value: arc_swap::ArcSwapOption::const_empty(),
+            source_type: ConfigSourceOrigin::RemoteConfig,
+        }
+    }
+
+    fn source(&self) -> ConfigSourceOrigin {
+        if self.override_value.load().is_some() {
+            self.source_type
+        } else {
+            self.config_item.source()
+        }
+    }
+
+    /// Replaces override value only if origin matches source_type
+    fn set_override_value(&self, value: T, source: ConfigSourceOrigin) {
+        if source == self.source_type {
+            self.override_value.store(Some(Arc::new(value)));
+        }
```

Comment:
If this is already called `set_override_value` why are we checking that the source is correct, and doing nothing if it's not?

I'm not sure we need the check, or if we need the check to protect against bugs it should return an error. Otherwise we're going to miss config updates

Reply:
My idea was to allow only updates for the source for which `ConfigItemWithOverride` was created for, if a `ConfigItemWithOverride` is created for RC updates I don't want to update it with a Code source...

I could log a warn if the value is not updated or return a `Result` to indicate the caller if the update went OK or not